src/integer_mod_q/mat_zq/inverse.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright © 2023 Marcel Luca Schmidt
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `inverse` function.
+
+use super::MatZq;
+use crate::{
+    integer::{MatZ, Z},
+    traits::{Concatenate, Gcd, GetNumRows},
+};
+use flint_sys::fmpz_mod_mat::fmpz_mod_mat_rref;
+
+impl MatZq {
+    /// Returns the inverse of the matrix if it exists (is square and
+    /// has a determinant unequal to zero) and `None` otherwise.
+    ///
+    /// Note that the modulus is assumed to be prime.
+    /// If it is not, it can happen that the function panics.
```

Comment:
This way, this function is really annoying to use. Will it panic or not when the modulus is not prime, nobody knows, but you will not be able to recover from it.
Also, I think that it will currently always panic in that case.

Suggestion: just allow prime moduli
Also, I don't think that Option is the right return type then. I think Result would fit better, because we can then insert into the error message why it did not work.

Reply:
It should be an option since we always have an option in inverse functions

Reply:
ok...